src/interfaces/chain.cpp

```diff
@@ -367,6 +367,13 @@ class ChainImpl : public Chain
     {
         return MakeUnique<RpcHandlerImpl>(command);
     }
+    void notifyMemoryPoolTransactions(Notifications& notifications) override
+    {
+        LOCK2(::cs_main, ::mempool.cs);
```

Comment:
Not sure how to solve the deadlock warning. Maybe take cs_wallet before both of these?

```
node1 2019-03-25T17:06:10.586193Z Leaving InitialBlockDownload (latching to false) 
 node1 2019-03-25T17:06:10.586301Z POTENTIAL DEADLOCK DETECTED 
 node1 2019-03-25T17:06:10.586314Z Previous lock order was: 
 node1 2019-03-25T17:06:10.586326Z  ::cs_main  interfaces/chain.cpp:372 
 node1 2019-03-25T17:06:10.586335Z  (1) ::mempool.cs  interfaces/chain.cpp:372 
 node1 2019-03-25T17:06:10.586350Z  cs_main  interfaces/chain.cpp:264 
 node1 2019-03-25T17:06:10.586359Z  (2) cs_wallet  wallet/wallet.cpp:1230 
 node1 2019-03-25T17:06:10.586386Z Current lock order is: 
 node1 2019-03-25T17:06:10.586397Z  cs_main  interfaces/chain.cpp:264 
 node1 2019-03-25T17:06:10.586406Z  (2) cs_wallet  wallet/wallet.cpp:1871 
 node1 2019-03-25T17:06:10.586419Z  (1) pool.cs  validation.cpp:580

Reply:
re: https://github.com/bitcoin/bitcoin/pull/15652#discussion_r268760261

> Not sure how to solve the deadlock warning. Maybe take cs_wallet before both of these?

It looks like you need to do the opposite and release cs_wallet before calling this.

Reply:
The lock order is `cs_main, cs_wallet, ::mempool.cs`, no?

Reply:
I can get the functional tests to pass with the diff

```diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index cb76461647..94e7a7ad53 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -367,9 +367,9 @@ public:
     {
         return MakeUnique<RpcHandlerImpl>(command);
     }
-    void notifyMemoryPoolTransactions(Notifications& notifications) override
+    void notifyMemoryPoolTransactions(Notifications& notifications, RecursiveMutex&cs_wallet) override
     {
-        LOCK2(::cs_main, ::mempool.cs);
+        LOCK2(::cs_main, cs_wallet);         LOCK(::mempool.cs);
         for (const CTxMemPoolEntry& entry : ::mempool.mapTx) {
             notifications.TransactionAddedToMempool(entry.GetSharedTx());
         }
diff --git a/src/interfaces/chain.h b/src/interfaces/chain.h
index 3fa63d83be..b53d170645 100644
--- a/src/interfaces/chain.h
+++ b/src/interfaces/chain.h
@@ -7,6 +7,7 @@
 
 #include <optional.h>               // For Optional and nullopt
 #include <primitives/transaction.h> // For CTransactionRef
+#include <sync.h> // For RecursiveMutex
 
 #include <memory>
 #include <stddef.h>
@@ -271,7 +272,7 @@ public:
     virtual std::unique_ptr<Handler> handleRpc(const CRPCCommand& command) = 0;
 
     //! Notify handler of all memory pool transactions.
-    virtual void notifyMemoryPoolTransactions(Notifications& notifications) = 0;
+    virtual void notifyMemoryPoolTransactions(Notifications& notifications,RecursiveMutex&cs_wallet) = 0;
 };
 
 //! Interface to let node manage chain clients (wallets, or maybe tools for
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 554ee62547..d72201679a 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -4404,7 +4404,7 @@ void CWallet::postInitProcess()
     ReacceptWalletTransactions();
 
     // Update wallet transactions with current mempool transactions.
-    m_chain.notifyMemoryPoolTransactions(*this);
+    m_chain.notifyMemoryPoolTransactions(*this, cs_wallet);
 }
 
 bool CWallet::BackupWallet(const std::string& strDest)